### PR TITLE
Add keyword advanced filter for Search

### PR DESCRIPTION
**Pull request by 289Adam289**
### Fixed Issues

$https://github.com/Expensify/App/issues/46027
PROPOSAL:

### Tests

- visit root `/search/filters`
- test keyword filter and whether navigation behaves correctly

- [x] Verify that no errors appear in the JS console

### QA Steps

Same steps as Tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/3ad5c154-214d-4102-b772-ebb63e494920

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/cc239692-a015-4d33-bd17-7ec01223f572

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/2d98295b-6022-46b7-87c3-d38c85e73d55

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/8776e7f2-509a-44cf-b04e-32d201ac55c6

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/8c130508-bdb9-434c-80c4-ea1b76b051fa

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/2ee7e74f-9c0d-40f4-ad62-b6e4e1244509

</details>

### CODE REVIEW

File: src/ROUTES.ts
Comment (Kicu) [inline src/ROUTES.ts:58]: remove empty line before this route

File: src/languages/es.ts
Comment (Kicu) [inline src/languages/es.ts:3674]: @luacmartins please verify

**Review by Kicu — APPROVED** [review]
Looks simple and legit 👍

**Comment by 289Adam289** [comment]
@luacmartins ready for review!

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by ikevin127** [comment]
> [!note]
> 1. The implementation only works as expected with **1 keyword**, if we add **2 or more** -> we get infinite loading.
> 2. 
##

⚠️ I completed the ✅ **PR Reviewer Checklist**, but I found one main issue with the functionality which manifests in 3 different ways which I see as possible blockers for this PR:

**Note**: The main issue has to do with using more than 1 keyword with different separators in between (comma, space).

* [ ] **1. Filters label issues and infinite loading when using comma separated keywords**

    **Notes:**
    1. This happens on all platforms.
    2. Notice the **Filters** labels and how there are 2 issues there:
    - each keyword appears separately as its own `keyword`
    - the `,` (comma) is seen as a separate keyword
    .

    <details>
    <summary>Videos / Screenshots</summary>
    
    https://github.com/user-attachments/assets/018aa00e-524f-48e7-b709-23b70d59a8bb
    
    | MacOS: Web |
    | ------------- |
    | Screenshot 2024-08-05 at 15 25 59 |

    </details>

##

* [ ] **2. Page not found when using semicolon separated keywords**

    **Notes:**
    1. This happens on all platforms.
    .

    <details>
    <summary>Video</summary>
    
    https://github.com/user-attachments/assets/8a47960f-b35b-4613-bea2-5905334263ee

    </details>

##

* [ ] **3. Infinite loading when using 2 or more space separated keywords**

    **Note:** This happens on all platforms.

    <details>
    <summary>Video</summary>
    
    https://github.com/user-attachments/assets/ca4b00bd-9c8f-4fe5-af3a-e930d5dc13f9

    </details>

##

> [!important]
> 1. I noticed the Filters category Capitalizing the keywords - I reported the issue before in [#46477 (comment)](https://github.com/Expensify/App/pull/46477#pullrequestreview-2211298801) which was resolved by @luacmartins in [#46477 (comment)](https://github.com/Expensify/App/pull/46477#pullrequestreview-2212587869) as "will be fixed in a follow-up PR". Just wanted to mention this again!
> 2. I think the infinite loading issues might be addressed in PR [#46804](https://github.com/Expensify/App/pull/46804) as I've reported them before in [#46566 (comment)](https://github.com/Expensify/App/pull/46566#issuecomment-2266188024) and the issue was addressed in [#46566 (comment)](https://github.com/Expensify/App/pull/46566#issuecomment-2268627463).

ℹ️ **Note:** I think it would be useful from a UI / UX POV to have a subtitle on the keyword input that mentions a certain separator can be used i case the user wants to have multiple keywords (cc @Expensify/design) and possible look at ways to make sure other separators / symbols won't show page not found or infinite loading when used.

##

@Kicu Please let me know if these are within the scope of this PR and if they should be addressed before moving forward. There are some conflicts as well. Once all of these are addressed, I will be able to 🟢 **Approve**.

**Comment by dubielzyk-expensify** [comment]
Would love a video/screenshot to see how this looks with some weird string examples like:
1. multiple words
2. symbols characters
3. emoji

**Comment by ikevin127** [comment]
@dubielzyk-expensify 👋 I dropped 3 videos in the https://github.com/Expensify/App/pull/46799#issuecomment-2270045899 above, are you now able to see them ?

They are related to the following scenarios:
- using comma separator
- using semicolon separator
- using space separator

Didn't try emojis yet.

**Comment by luacmartins** [comment]
FYI the keyword filter is not yet supported in the backend and is throwing [this error](https://github.com/Expensify/Expensify/issues/416639). I'm working on a PR to fix that.

**Comment by Kicu** [comment]
- with the newest `main` merged (good job @289Adam289) we should no longer be seeing multiple `keyword:xxx keyword:yyy` within the results page query that we display to the user - this was a bug that is solved. For every multivalue filter we want to see `category:XXX,YYY keyword:1,2,3` etc.

That being said...
I believe it is worth for someone from Expensify team to describe exactly what should happen in these cases:
 - user types several words separated with space into keyword (ex `Las Vegas bill`)
 - user types several words separated with comma into keyword (ex `Las,Vegas,bill`)
 - user types words with mix of spaces and commas

For each of these what should be the final value of `keyword` filter that is being sent to backend? @luacmartins

**Comment by luacmartins** [comment]
[PR](https://github.com/Expensify/Auth/pull/11905) to handle keyword filter in review

**Comment by luacmartins** [comment]
I think for now, we should treat space and commas as separators and do an `OR` search on each word, i.e.:

>user types several words separated with space into keyword input (ex Las Vegas bill)

would search for `Las` OR `Vegas` OR `bill`

>user types several words separated with comma into keyword input (ex Las,Vegas,bill)

would search for `Las` OR `Vegas` OR `bill`

>user types words with mix of spaces and commas  Las, Vegas, casino bill

would search for `Las` OR `Vegas` OR `casino` OR `bill`

The keyword filter is one that we'll have to refine over time, e.g. drop common words like `a`, `of`, etc, fix typos, synonyms, etc but the behavior above should be a good starting point. cc @JmillsExpensify @trjExpensify for thoughts

**Comment by shawnborton** [comment]
From @ikevin127 's third example, when the user types in `CAR RON TESLA`:
CleanShot 2024-08-06 at 08 53 40@2x

The filter title spits it out as three separate keywords:
CleanShot 2024-08-06 at 08 53 26@2x

Why is that? Even if we are searching three separate words, it feels weird to the user that the string got chopped up. Can we just use `keywords:CAR RON TESLA` in the title?

**Comment by ikevin127** [comment]
> Can we just use keywords:CAR RON TESLA in the title?

✅ That was fixed after another PR was recently merged and is now fixed on this one as well after sync w/ main:

Screenshot 2024-08-06 at 12 57 39

##

In summary, to callback to the issues found (https://github.com/Expensify/App/pull/46799#issuecomment-2270045899), what's needed here in order to move forward with **Approval** and merge:
1. From [#46799 (comment)](https://github.com/Expensify/App/pull/46799#issuecomment-2270804297) -> we're looking at deciding what's the best way to go for the first iteration of this (to be improved in future follow-ups) and based on decision, changes will be pushed to this PR.
2. Regarding the infinite loading, NAB I think for now (?!) as **Carlos** [mentioned](https://github.com/Expensify/App/pull/46799#issuecomment-2270730763) that the keyword search is not yet supported in BE so it's expected to get the infinite loading when using more than 1 keyword ? @luacmartins 
3. Which leaves the issue number (2.) **Page not found when using semicolon separated keywords** to be addressed.

**Comment by shawnborton** [comment]
Hmm I still think, as a user, I would expect to see `keywords:CAR RON TESLA` and not `keyword:CAR,RON,TESLA`

**Comment by dubielzyk-expensify** [comment]
Agree with you Shawn. Cause it's be weird if you add comma in your string, then see double commas.

**Comment by luacmartins** [comment]
I agree. I think we need to change the grammar to combine all filters with the same key.

**Comment by 289Adam289** [comment]
@ikevin127 **2. Page not found using semicolon** problem comes up in multiple filters. After discussing it with team and @luacmartins we agreed to handle it as a follow up. Carlos have you had a chance to create a issue for that? If not would you mind? Thanks!

**Comment by trjExpensify** [comment]
>Why is that? Even if we are searching three separate words, it feels weird to the user that the string got chopped up. Can we just use keywords:CAR RON TESLA in the title?

I agree with Shawn here. 👍

**Comment by luacmartins** [comment]
>@ikevin127 2. Page not found using semicolon problem comes up in multiple filters. After discussing it with team and @luacmartins we agreed to handle it as a follow up. Carlos have you had a chance to create a issue for that? If not would you mind? Thanks!

@289Adam289 created issue [here](https://github.com/Expensify/App/issues/46993)

**Comment by luacmartins** [comment]
@289Adam289 it seems like we still need to address some comments on how we display the filters to the user.

**Comment by shawnborton** [comment]
Much better!

**Review by luacmartins — APPROVED** [review]
The PR is looking good. There are a few issues with the parser grammar that make some of the results weird when using special characters, e.g. `,`, `;`, etc I think those can be addressed in a follow up though.

I'm gonna merge this PR and create a follow up issue to update the parser grammar so it handles the special character cases better.

**Comment by luacmartins** [comment]
Created issue to update parser grammar [here](https://github.com/Expensify/App/issues/47095)

**Comment by Beamanator** [comment]
FYI I believe this was deployed to prod yesterday, from this checklist - https://github.com/Expensify/App/issues/47219

**Pull request merged**
